src/mip/utils/timestamp.hpp

```diff
@@ -0,0 +1,222 @@
+        template<typename D> D getTimestamp();
+        Nanoseconds getTimestamp();
+
+    private:
+        Nanoseconds timestamp{0}; 
+    };
+
+    /// Returns time elapsed since the start of the current week of the timestamp.
+    ///
+    /// Example usage:
+    ///     // Get time of week in the same duration unit as the input timestamp.
+    ///     Nanoseconds timestamp{2000000000};
+    ///     Nanoseconds same_out = getTimeOfWeek(timestamp);
+    ///
+    ///     // Get time of week in a different duration unit from the input timestamp.
+    ///     Nanoseconds timestamp{2000000000};
+    ///     Seconds diff_out = getTimeOfWeek(duration_cast<Seconds>(timestamp));
+    template<typename D> 
+    D getTimeOfWeek(const D &timestamp);
+
+    /// Returns whether two timestamps have diverged from each other.
+    ///
+    /// Timestamps are considered diverged if they differ by one or more units of the 
+    /// comparison duration.
+    /// 
+    /// Example usage:
+    ///     // Timestamps are different by one second.
+    ///     Seconds timestamp1{2};
+    ///     Seconds timestamp2{1};
+    ///     bool changed1 = timeChanged<Seconds>(timestamp1, timestamp2);
+    ///     bool changed2 = timeChanged<Seconds>(timestamp2, timestamp1);
+    ///     // ---> changed1 is true.
+    ///     // ---> changed2 is also true. Comparison works both ways!
+    ///
+    ///     // Timestamps are different by more than one nanosecond.
+    ///     Seconds timestamp1{2}; // 2,000,000,000 nanoseconds
+    ///     Seconds timestamp2{1}; // 1,000,000,000 nanoseconds
+    ///     bool changed = timeChanged<Nanoseconds>(timestamp1, timestamp2); 
+    ///     // ---> changed is true. 
+    ///     // ---> Comparison is done in nanoseconds instead of seconds.
+    ///
+    ///     // Timestamps aren't different by a week or more.
+    ///     Seconds timestamp1{604800}; // 1 week
+    ///     Seconds timestamp2{604799}; // 1 second less than a week
+    ///     bool changed = timeChanged<Weeks>(timestamp1, timestamp2); 
+    ///     // ---> changed is false. 
+    ///     // ---> Comparison is done in weeks.
+    template<typename DCompare, typename D1, typename D2> 
+    bool timeChanged(const D1 &timestamp1, const D2 &timestamp2);
+
+    /// Casts the timestamp duration to the given arithmetic type.
+    ///
+    /// Optionally performs a duration cast on the timestamp duration before casting 
+    /// to the arithmetic type.
+    ///
+    /// Example usage:
+    ///     // Cast seconds to int.
+    ///     Seconds timestamp{123};
+    ///     int time_casted = castTime<int>(timestamp);    
+    ///     ---> time_casted = 123.
+    ///
+    ///     // Cast nanoseconds to seconds to int.
+    ///     Nanoseconds timestamp{1000000000}; // This equals one second.
+    ///     int time_casted = castTime<int, Seconds>(timestamp);
+    ///     ---> time_casted = 1.
+    template<typename T, typename DCast, typename DIn> 
+    T castTime(const DIn &timestamp);
```

Comment:
It would evaluate to 1.0 because the conversion is done for the time duration unit first. So the seconds would first be converted to weeks, which would be Weeks(1). Then the weeks to float conversion would occur, which would be 1.0